p2p/discover/metrics.go

```diff
@@ -0,0 +1,65 @@
+// Copyright 2023 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package discover
+
+import (
+	"net"
+
+	"github.com/ethereum/go-ethereum/metrics"
+)
+
+const (
+	moduleName = "discover"
+	// ingressMeterName is the prefix of the per-packet inbound metrics.
+	ingressMeterName = moduleName + "/ingress"
+
+	// egressMeterName is the prefix of the per-packet outbound metrics.
+	egressMeterName = moduleName + "/egress"
+)
+
+var (
+	ingressTrafficMeter = metrics.NewRegisteredMeter(ingressMeterName, nil)
+	egressTrafficMeter  = metrics.NewRegisteredMeter(egressMeterName, nil)
+)
+
+// meteredConn is a wrapper around a net.UDPConn that meters both the
+// inbound and outbound network traffic.
+type meteredUdpConn struct {
+	UDPConn
+}
+
+func newMeteredConn(conn UDPConn) UDPConn {
+	// Short circuit if metrics are disabled
+	if !metrics.Enabled {
```

Comment:
I am not fully sure that we should use `metrics.Enabled` or `metrics.EnabledExpensive`.

The former one will be enabled once the metric system is turned on. The latter one is more for debug purpose which might  impact runtime performance.